### PR TITLE
Set always a default timeout on knative build to avoid hanging builds

### PR DESCRIPTION
**Pull request by ccojocar**
Define a new option to the build controller which can be use to set the default timeout.

### CODE REVIEW

File: prow/cmd/build/main.go [hunk 1]
```diff
@@ -70,6 +71,7 @@ func (o *options) parse(flags *flag.FlagSet, args []string) error {
 	flags.StringVar(&o.buildCluster, "build-cluster", "", "Path to file containing a YAML-marshalled kube.Cluster object. If empty, uses the local cluster.")
 	flags.StringVar(&o.cert, "tls-cert-file", "", "Path to x509 certificate for HTTPS")
 	flags.StringVar(&o.privateKey, "tls-private-key-file", "", "Path to matching x509 private key.")
+	flags.DurationVar(&o.buildTimeout, "build-timeout", 1*time.Hour, "Default timeout for build. Only used if not timeout is defined in the prow job")
```
Comment (stevekuznetsov) [inline prow/cmd/build/main.go]: On Pod jobs we have timeouts per-job, configurable default in Plank config and an ability to override on a per-job level. That may be a better approach here?

Comment (krzyzacy) [reply prow/cmd/build/main.go]: hummmmm the default podutils timeout should apply to the build jobs as well? no?

Comment (ccojocar) [reply prow/cmd/build/main.go]: The issue is when the timeout in the prow job is set to 0 (default). The same value is also set into the build resulting in an infinite timeout. Also the same situation is when the job decoration is not set.

This ensure basically that always there is a timeout, but always the timeout from the job or build has precedence over this if is configured.

Comment (stevekuznetsov) [reply prow/cmd/build/main.go]: Why is that a bad timeout value? Is it not just as harmful on Pods as it is on Builds? If a user wants a global default, why wouldn't they set one?

Comment (ccojocar) [reply prow/cmd/build/main.go]: The user does not have always control over this value. The prow jobs are created by `tide` with no timeout set. See https://github.com/kubernetes/test-infra/blob/c568296819c5e1b355d8e574284dada33b5c8241/prow/config/config.go#L1198 and https://github.com/kubernetes/test-infra/blob/c568296819c5e1b355d8e574284dada33b5c8241/prow/pjutil/pjutil.go#L108

Comment (stevekuznetsov) [reply prow/cmd/build/main.go]: `tide` is just using the configuration that was created by users? In the `PresubmitSpec` function you link to, the `p config.Presubmit` is the configuration on disk for the job, which is written by users.

Comment (ccojocar) [reply prow/cmd/build/main.go]: That's right, but imagine that you have 50+ jobs defined in the configuration but some of them are missing the timeout. The intend of this timeout is to increase the reliability of the build. We had this situation in Jenkins X.

Comment (stevekuznetsov) [reply prow/cmd/build/main.go]: You can set the default timeout in the `plank` configuration, though, for all jobs. Then only jobs that explicitly override will get anything else.

Comment (ccojocar) [reply prow/cmd/build/main.go]: Which timeout are you referring to? `PodPendingTimeout`? As far as I can see in the code, this timeout is only used for pods created by plank for prow jobs with `kubernetes` agent, but not `kanative-build` agent.  

This default timeout is only for `knative-build` controller. I can also add the vaulue in the knative build controller config.

**Comment by stevekuznetsov** [comment]
This seems like it will fix part of #7152 

I think we should continue to have:
 - controller-specific timeouts configured through `config.yaml`
 - a mechanism to set timeouts per-job and overwrite the above
 - perhaps to really finish #7152 something at a global level in `sinker`?

**Comment by ccojocar** [comment]
@stevekuznetsov Should I define a default timeout also in the build config? `plank` already has one.  

The timeout set per job overwrite currently this timeout. 

Please could you provide a bit more details about what you mean with something at global level `sinker`?

**Comment by stevekuznetsov** [comment]
If we want parallel config for the controllers, a controller-wide timeout for `build` seems appropriate, yes. We should plumb it through config, not a flag, though.

`sinker` is the component that enforces Pod deletion -- we could do a controller-agnostic timeout there for any pods created by Prow.

@cjwagner @fejta would like to weigh in I imagine, maybe you should bring this to the SIG meeting today to hash out what we want the config to look like

File: prow/cmd/build/main.go (hunk 1 above)
Comment (fejta) [reply prow/cmd/build/main.go]: > On Pod jobs we have timeouts per-job, configurable default in Plank config and an ability to override on a per-job level. That may be a better approach here?

Yes, please reuse these values.

> PodPendingTimeout? As far as I can see in the code, this timeout is only used for pods created by plank for prow jobs with kubernetes agent, but not kanative-build agent.

Please teach knative-build to use this value, just like we taught it to use PodNamespace. Let's not create a separate value.

**Comment by stevekuznetsov** [comment]
> We don't

What do we want? Let's be prescriptive here so we can just implement what we need.

File: prow/cmd/build/main.go (hunk 1 above)
Comment (stevekuznetsov) [reply prow/cmd/build/main.go]: Yes, when you move it we need to continue loading the old value

**Comment by fejta** [comment]
We want plank and build to have the same behavior, configured via the same config values.

If the job sets an explicit `timeout`, use this value https://github.com/kubernetes/test-infra/blob/b754ac8b38be8402d7150a94f09151a246ca4dce/prow/apis/prowjobs/v1/types.go#L155-L158

https://github.com/kubernetes/test-infra/blob/b754ac8b38be8402d7150a94f09151a246ca4dce/prow/config/jobs.go#L403

https://github.com/kubernetes/test-infra/blob/b754ac8b38be8402d7150a94f09151a246ca4dce/prow/config/jobs.go#L108

https://github.com/kubernetes/test-infra/blob/b754ac8b38be8402d7150a94f09151a246ca4dce/prow/config/jobs.go#L113

Otherwise, use the `plank.default_decoration_config.timeout` value in config.yaml

https://github.com/kubernetes/test-infra/blob/master/prow/config/config.go#L68-L70

https://github.com/kubernetes/test-infra/blob/master/prow/config/config.go#L192

**Comment by fejta** [comment]
So we may not yet have this feature in plank... IMO we have a Timeout value for the test container, and a `PodPendingTimeout` for the controller.

Neither of these are really what we want:
* `Timeout` sets a deadline on containers once they starts... but that requires a running container, which does not always happen
* `PodPendingTimeout` is only relevant when the pod hasn't yet started and is in a state of pending.

I think it might be useful to create a deadline value for jobs and add a default deadline for which this is unconfigured at the job level, and teach both plank and build to follow it.

**Comment by stevekuznetsov** [comment]
I guess also -- do we want this as a top-level thing that Plank and Build both read, or do you want Build to read Plank config...

File: prow/cmd/build/main.go (hunk 1 above)
Comment (ccojocar) [reply prow/cmd/build/main.go]: I removed the flag and define a config value.

**Comment by ccojocar** [comment]
>I think it might be useful to create a deadline value for jobs and add a default deadline for which this is unconfigured at the job level, and teach both plank and build to follow it.

@fejta @stevekuznetsov I defined a global deadline in the prow config, which defaults to 24 hours if is not set in order to avoid a timeout of infinity.

I changed the build controller to use the value from the config instead of the flag when there is not timeout set per job.   

I think, we can add this feature to `Plank` in a follow up PR.

File: prow/cmd/build/controller_test.go [hunk 2]
```diff
@@ -1212,10 +1224,11 @@ func TestDecorateBuild_Timeout(t *testing.T) {
 		expected         *time.Duration
 	}{
 		{
-			name: "do not change timeout when decorated unset",
+			name:     "set the default timeout when decoration is unset",
+			expected: &defaultTimeout,
 		},
 		{
-			name:         "do not change set timeout (decoration unset)",
```
Comment (fejta) [inline prow/cmd/build/controller_test.go]: Why is this change necessary?

File: prow/cmd/build/controller_test.go [hunk 3]
```diff
@@ -1230,6 +1243,11 @@ func TestDecorateBuild_Timeout(t *testing.T) {
 			decoratedTimeout: &b,
 			expected:         &b,
 		},
+		{
+			name:             "set the default timeout when unset and decorated set with infinite timeout",
```
Comment (fejta) [inline prow/cmd/build/controller_test.go]: Seems more correct here to say that the decorated timeout is zero.

File: prow/cmd/build/admission.go [hunk 4]
```diff
@@ -43,9 +43,18 @@ var (
 )
 
 func init() {
-	prowjobscheme.AddToScheme(vscheme)
-	admissionapi.AddToScheme(vscheme)
-	admissionregistrationv1beta1.AddToScheme(vscheme)
+	err := prowjobscheme.AddToScheme(vscheme)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/cmd/build/admission.go]: fixed

File: prow/cmd/build/controller.go [hunk 5]
```diff
@@ -114,30 +125,34 @@ func (c *controller) hasSynced() bool {
 	return true // Everyone is synced
 }
 
-func newController(kc kubernetes.Interface, pjc prowjobset.Interface, pji prowjobinfov1.ProwJobInformer, buildConfigs map[string]buildConfig, totURL string, prowConfig config.Getter, rl workqueue.RateLimitingInterface) *controller {
+func newController(opts *controllerOptions) *controller {
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/cmd/build/controller.go]: fixed by removing the pointer

File: prow/cmd/build/controller.go [hunk 6]
```diff
@@ -649,13 +661,25 @@ func injectedSteps(encodedJobSpec string, dc prowjobv1.DecorationConfig, injecte
 	return []coreapi.Container{placer, *initUpload}, sidecar, &gcsVol, nil
 }
 
+func injectTimeout(spec *buildv1alpha1.BuildSpec, dc *prowjobv1.DecorationConfig, defaultTimeout time.Duration) {
+	if dc != nil {
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/cmd/build/controller.go]: I refactored the code as suggested in a separate function.

File: prow/cmd/build/main.go [hunk 7]
```diff
@@ -71,7 +72,10 @@ func (o *options) parse(flags *flag.FlagSet, args []string) error {
 	flags.StringVar(&o.buildCluster, "build-cluster", "", "Path to file containing a YAML-marshalled kube.Cluster object. If empty, uses the local cluster.")
 	flags.StringVar(&o.cert, "tls-cert-file", "", "Path to x509 certificate for HTTPS")
 	flags.StringVar(&o.privateKey, "tls-private-key-file", "", "Path to matching x509 private key.")
-	flags.Parse(args)
+	err := flags.Parse(args)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/cmd/build/main.go]: fixed

File: prow/config/config.go [hunk 8]
```diff
@@ -1043,6 +1047,11 @@ func parseProwConfig(c *Config) error {
 	}
 	logrus.SetLevel(lvl)
 
+	// Avoid using a job timeout of infinity by setting the default value to 24 hours
+	if c.DefaultJobTiemout == 0 {
+		c.DefaultJobTiemout = time.Hour * 24
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/config/config.go]: extracted in a const

File: prow/cmd/build/controller_test.go [hunk 9]
```diff
@@ -1201,7 +1211,9 @@ func TestInjectedSteps(t *testing.T) {
 	}
 }
 
-func TestDecorateBuild_Timeout(t *testing.T) {
+func TestInjectTimeout(t *testing.T) {
+	defaultTimeout := 1 * time.Hour
+	var infiniteTimeout time.Duration = 0
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/cmd/build/controller_test.go]: a) removed the variable initialisation
b) It is required because the tested method expects a pointer.

File: prow/cmd/build/controller.go [hunk 10]
```diff
@@ -114,30 +125,34 @@ func (c *controller) hasSynced() bool {
 	return true // Everyone is synced
 }
 
-func newController(kc kubernetes.Interface, pjc prowjobset.Interface, pji prowjobinfov1.ProwJobInformer, buildConfigs map[string]buildConfig, totURL string, prowConfig config.Getter, rl workqueue.RateLimitingInterface) *controller {
+func newController(opts *controllerOptions) *controller {
 	// Log to events
-	prowjobscheme.AddToScheme(scheme.Scheme)
+	err := prowjobscheme.AddToScheme(scheme.Scheme)
+	if err != nil {
+		logrus.Fatalf("Add prowjob scheme: %v", err)
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/cmd/build/controller.go]: fixed

File: prow/cmd/build/controller_test.go (hunk 2 above)
Comment (ccojocar) [reply prow/cmd/build/controller_test.go]: It was unintentional. Restored the previous test name.

**Comment by ccojocar** [comment]
@fejta Thanks for review! I addressed all your comments.

File: prow/config/config.go [hunk 11]
```diff
@@ -118,6 +118,10 @@ type ProwConfig struct {
 	// found, or have another generic issue. The default that will be used if this is not set
 	// is: https://github.com/kubernetes/test-infra/issues
 	StatusErrorLink string `json:"status_error_link,omitempty"`
+
+	// DefaultJobTiemout this is default deadline for prow jobs. This value is used when
+	// no timeout is configured at the job level. This value is set to 24 hours.
+	DefaultJobTiemout time.Duration `json:"default_job_tiemout,omitempty"`
```
[the remark this answers is not in the document]
Comment (fejta) [reply prow/config/config.go]: Did you forget to push a commit? These all seem unaddressed

File: prow/cmd/build/admission.go [hunk 12]
```diff
@@ -43,9 +43,18 @@ var (
 )
 
 func init() {
-	prowjobscheme.AddToScheme(vscheme)
-	admissionapi.AddToScheme(vscheme)
-	admissionregistrationv1beta1.AddToScheme(vscheme)
+	err := prowjobscheme.AddToScheme(vscheme)
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/cmd/build/admission.go]: I tried to fix some lint warnings, but I will move these changes in a different PR.

File: prow/cmd/build/main.go [hunk 13]
```diff
@@ -115,7 +119,7 @@ func newBuildConfig(cfg rest.Config, stop chan struct{}) (*buildConfig, error) {
 
 	// Ensure the knative-build CRD is deployed
 	// TODO(fejta): probably a better way to do this
-	_, err = bc.Build().Builds("").List(metav1.ListOptions{Limit: 1})
+	_, err = bc.BuildV1alpha1().Builds("").List(metav1.ListOptions{Limit: 1})
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/cmd/build/main.go]: Tried to fix a lint warning. It seems that `Build()` is marked as deprecated to encourage to use the versioned method. I will revert it and send a separate PR with all the lint fixes.

File: prow/config/config.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/config/config.go]: Yeah, should be fixed now.

File: prow/cmd/build/main.go [hunk 14]
```diff
@@ -190,7 +194,17 @@ func main() {
...
-	controller := newController(kc, pjc, pjif.Prow().V1().ProwJobs(), buildConfigs, o.totURL, configAgent.Config, kube.RateLimiter(controllerName))
+	opts := &controllerOptions{
+		kc:           kc,
+		pjc:          pjc,
+		pji:          pjif.Prow().V1().ProwJobs(),
+		buildConfigs: buildConfigs,
+		totURL:       o.totURL,
+		prowConfig:   configAgent.Config,
+		rl:           kube.RateLimiter(controllerName),
+		buildTimeout: configAgent.Config().DefaultJobTiemout,
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/cmd/build/main.go]: The value is loaded dynamically now from config each time `defaultBuildTImeout() gets called.

File: prow/config/config.go [hunk 15]
```diff
@@ -1043,6 +1047,11 @@ func parseProwConfig(c *Config) error {
 	}
 	logrus.SetLevel(lvl)
 
+	// Avoid using a job timeout of infinity by setting the default value to 24 hours
+	if c.DefaultJobTiemout == 0 {
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/config/config.go]: Can you point me to an example which shows how to add such a flag in the config? Should I just add a bool flag in the config?

**Comment by fejta** [comment]
Thanks!
